### DISCUSSION TITLE
Ride feeling

### DISCUSSION THREAD

**Discussion by VincentD75** [JuergenLeber/SHIFTR#10] [General]
Hello Juergen,

Built is done and eveything looks ok.

When riding, trainer difficulty is changing from low to high and high to low even on flat road. 

I am using wifi connection. 

Any idea what could be the issue ?

**Comment by iacchos-fr** [comment]
I assume you are in target mode (default). Change to "basic resistance"...

**Reply by VincentD75** [reply]
Indeed, i will try with basic resistance. Thanks for the input.

**Reply by VincentD75** [reply]
I confirm it is working like a charm in basic resistance mode.

**Reply by iacchos-fr** [reply]
Great ! @JuergenLeber : not sure to understand difference between "Track Resistance" and "Basic Resistance" (considering the Neo 2T) ??

**Reply by JuergenLeber** [reply]
Cool, great to have another user @VincentD75! I just did a new release 1.2.0 last week which changes the calculation method in target power mode. Did you also try that?

Regarding the three modes: The TACX Neo 2T (and all the other trainers that support FE-C via BLE) of course doesn't support virtual shifting but instead four different modes to set the resistance:

- Basic Resistance = Here I can set a value between 0% and 100% (in 0.5% steps) which resistance the trainer should apply. The maximum resistance will be requested beforehand and then I try to calculate a good value regarding on gear, weight, incline and speed.
- Target Power = Here I can set a value between 0W and 4000W (in 0.25W steps). This ist the mode that is also being used in ERG modes when you're e.g. doing workouts. I also try to calculate a good value here based on gear, weight, incline and speed.
- Wind Resistance = This mode isn't being used in SHIFTR. I could set a wind speed and it would set the corresponding resistance then but that doesn't help us here :-)
- Track Resistance = Here I can set the grade of the simulated track from -200% to +200% (in 0.01% steps). Also here I try to calculate a good value regarding on gear, weight, incline and speed.

The calculations are described in the README.md of the project but none of them will probably be 100% correct. That's why I need your feedback and testing to make those calculations even better and/or find the perfect "mode" and setting. My favorite is the "Target Power" mode after my recent changes but yours can be different. So please give me feedback then I can do changes accordingly!

As far as I got it you're both using "Basic Resistance" with a Neo 2T currently, right? Are the gears all okay or too hard/soft? I could additionally add another factor to do fine-tuning here without having to misuse the chainring/cog teeth setting here. What do you think?

**Reply by iacchos-fr** [reply]
Thanks for this summarize (easier to get the point than the calculation in the readme file 😀)

Yes I choose basic mode. However, the sensations are a little ‘mixed’.
My 1st difficulty is that I can feel the difference in gradient very strongly. With the "real shifting", I could climb slopes of up to 3% (and down to around 3%), without changing gear... With the SHIFTR in gear 12 or 13 on the flat, I have to go down to 5 or 6 otherwise the effort is far too great!
This morning I tried again, setting the Neo2T's max power in Zwift to 75%. It's better, it seems smoother. I did a test in Track resistance mode, but it didn't go far enough to make any difference....

For me, that's the main difficulty: I have to play a lot with the different gears as soon as the gradient changes, much more than with real shifting!


NB: I didn't understand the point of the power taget code: it's similar to ERG mode (for a workout) and therefore in ride simulation, I don't understand?

**Reply by JuergenLeber** [reply]
The "Target Power" mode is not similar to the ERG mode - It IS the ERG mode used in simulation mode and updated up to four times a second to match the current incline and gear setting. That's also the downside of this mode as with a lower cadence the force will be higher and needs some fractions of a second to adjust.

In all three modes I just do different calculations that will then match (or that I'm trying to match) the existing FE-C resistance/power setting to adjust the trainer. 

The parameters I receive and use from Zwift are: Weight (bike+user), grade and gear ratio

The parameters I receive and use from the trainer are: Cadence, power, maximum resistance

Taking those parameters I can do different calculations and even calculate other things like speed (based on cadence, gear ratio and wheel diameter). 

What I then do is to calculate the needed force (based on weight and incline) to move the bike which I set as the resistance in "Basic Resistance" mode or in combination with the speed as a target power in "Target Power" mode. This calculation is updated up to four times a second and an update is sent to the trainer. In "Track Resistance" mode I basically do the same thing but as it needs a grade/slope value I calculate back how much grade/slope would equal the needed force. So probably not the best mode to start trying.

Please don't change any parameters in Zwift as they can't fix the calculations and make comparisons between our approaches harder.

So let's focus on the "Basic Resistance" mode for now. That's the function "Calculations::calculateFECResistancePercentageValue" in "Calculations.cpp" file:

- First I calculate the speed based on cadence, wheelDiameter and the gear ratio that is sent from Zwift. For example with a cadence of 80 and Zwift gear 8 (ratio 1.68) = 4.92 m/s
- Next I calculate the total force needed based on weight (bike+user), grade and the speed from above. For example with 85kg user and 15kg bike (yes, heavy bike but easier to calculate with ;-) ) on a track with 1% incline = Gravity force 9.8N, rolling force 4.0N and drag force 6.2N in sum = 20N
- Next I calculate the relative gear ratio based on what we really have (e.g. 34T chainring and 14T cog = 2.43) and what Zwift wants (in our case gear 8 = 1.68) which results to: 0.69
- Last I take the force calculated in step 2 and correct it up or downwards related to the relative gear ratio. In our example it would lead to a force of around 13.9N
- Then - after asking the trainer for its maximum resistance - a value is set representing this 13.9N as accurate as possible in a range of 0-100% resistance.

And probably here we need to adjust some things. 

BTW: I never do any changes to the power that is transmitted to Zwift. Meaning that whatever we calculate and set in resistance the output power won't be faked or modified in any way and thus always be correct.

**Reply by iacchos-fr** [reply]
OK clear. Thanks for the precisions 
But why is it that when I change the maximum power of the smart trainer in Zwift, I get a ‘softer’ feeling (i.e. less impact on small inclines)?  I must admit than I am far from the "maximum resistance" of a Neo2T , and probably use only the first 1/4 of max power (0-600w)... 
I'm intrigued to understand the impact of changing (in Zwift) the power max I tested! Theoretically this change alters the result of the last step in your description above?

**Reply by JuergenLeber** [reply]
I need to recheck what Zwift does here but I just think they calculate the incline smaller in this case. That then also affects the SHIFTR calculation - but only one part of it. So I'd recommend to keep this at default 100% and we'll work on the calculation. 

I'm doing a ride right now in "Basic Resistance" mode and will recheck the resistance feeling. Probably I'll then come up with with another release by tomorrow!

**Reply by iacchos-fr** [reply]
Ah yes you're right : this option impacts the incline **only** (and as far as I remember, the incline of the downhill runs is always 2 times less than what is displayed...). So : smaller max power = smoother slope = smoother feel.
But in view of my ride this morning, I really think we need to change the gears a lot in virtual sharing...
I'll give it a try in target power mode...

**Reply by JuergenLeber** [reply]
No - we're not changing the gears (the ratio comes from Zwift) but we're changing the calculation. I didn't do a lot of bike trainings besides some workouts the last weeks as I'm currently training for a better running pace but I just tested it and I think I know what you mean. On small incline changes the reaction that is needed on the gears is too much. 

As all modes use the same basic calculations you don't need to try in target power mode. It may feel a bit better (because it can be adjusted way better than only 0-100%) but the base calculation has an issue that I'll need to fix. Or at least to change and then we'll try :-)

I'll come back with a new version and then both of you @VincentD75 and @iacchos-fr have to try please.

**Reply by iacchos-fr** [reply]
> On small incline changes the reaction that is needed on the gears is too much.

Yes, you get the point (and I almost never do workouts, only free rides or events :-) !!

Let's say that for the moment, the max power option in Zwift is a workaround, but I can't wait to try out a new calculation formula :-)  Of course, I'll give you feedback

Thanks again for the job :-)

**Reply by VincentD75** [reply]
Hello,

Just to share my feedback which is similar. Incline and decline seems to have too much impact on resistance value.  

Regards

**Reply by JuergenLeber** [reply]
Okay, give me a few days and I'll come up with a new version. Just did a test 5 minutes ago but that wasn't what I expected so it needs more fine tuning.

**Reply by iacchos-fr** [reply]
Hello @JuergenLeber 
Have you managed to make the expected changes? Thanks in advance

**Reply by JuergenLeber** [reply]
Hi @iacchos-fr!
I'm still on it but simply had no time yet (x-mas and so on 😃). Expect something for the very early next year!

**Reply by iacchos-fr** [reply]
Hi
No worries, enjoy the moment !
Happy new year :-)

**Reply by JuergenLeber** [reply]
@iacchos-fr Happy new year to you, too! I'm back from vacation and can now move on here ☺️

**Reply by iacchos-fr** [reply]
<hello @JuergenLeber 
Have you had the opportunity to make any adjustments to provide more smoothness on hilly roads?
Thanks

**Reply by JuergenLeber** [reply]
Hi @iacchos-fr !
Yes, I did a lot of tests with different methods the last weeks and I think I found a good solution. Will release it in a few days and I'm looking forward to your test results then!

**Comment by syahvian** [comment]
Hi @JuergenLeber, thanks for this awesome project and sharing it.

Related to the ride feeling, how about this https://github.com/cagnulein/qdomyos-zwift/discussions/2671, is it possible or not for the FE-C ecosystem? it is specific to the Wahoo ecosystem, but in the ANT+ specs page 55 (0x37), we can change the wheel diameter configuration too.

Regards.

**Reply by JuergenLeber** [reply]
Hi @syahvian!

That was my very first approach to just change the wheel diameter but unfortunately the TACX trainers don't care about that setting at all. I'm currently working on a solution that will probably do the trick. Additionally I'm experimenting with "smoothening" the hard incline changes e.g. from 0 to 8% in small steps instead of directly applying 8%. Do you already have the hardware built? I'd be happy to have another tester!

**Comment by JuergenLeber** [comment]
I just released a new version (1.3.0, see [Releases](https://github.com/JuergenLeber/SHIFTR/releases)) with completely updated calculations!

After the update make sure you're using the "Basic Resistance" mode first as this will probably have the best results as it is less dependent on the speed and cadence and so should provide better results. But also try the other modes and most important: Give me feedback. 

I updated the project to be able to change and check the calculations without having to flash it to the microcontroller first and so changed can be made much easier now. Looking forward to your tests!

**Reply by iacchos-fr** [reply]
Thank you very much @JuergenLeber for this work, I'm impressed! Unfortunately, I'm not sure I'll have much time for testing as I'm leaving in a few days for a 10-day vacation :-) I'll be sure to give you feedback as soon as I can!

**Reply by JuergenLeber** [reply]
Enjoy your vacation, feedback isn't urgent!

**Comment by dnks23** [comment]
Looping myself in here. I had some time for a short test ride with my setup including Zwift Ride, Tacx Neo 2T and Zwift on iPad, basic resistance mode.

So far I really liked the experience, so again thank you for the great work @JuergenLeber !
My first feedback would be that, personally, I had the feeling that the gears came in a bit too smooth and with a bit of delay compared to an outdoor ride where gears kick in immediately when shifting. This was basically my first ride with virtual shifting so I'm not sure what exactly can be configured as part of SHIFTR and to what extend things are simply related to the virtual shifting protocol of Zwift...

What are your thoughts on this? I'll try to test the other modes as well soon to see differences...

**Reply by JuergenLeber** [reply]
Thank you for testing! There will always be a small "delay" as the trainer itself doesn't apply the new force immediately. But the goal was to make it smoother and that seems to have worked :-) The "Basic Resistance" mode is the one that works best in my opinion but please also try the others, too.
